### PR TITLE
Fix #22: improve the way to check json_decode result

### LINKED ISSUE #22: json decoding is not handling nicely
output of json_decode is not handled, thus in case of any json-violation, we got following error:
```
Type error: Argument 1 passed to ComposerRequireChecker\Cli\Options::__construct() must be of the type array, null given, called in ...
```

### PR DESCRIPTION
**Pull request by villfa**
Fix for #22.

In case of json violation `json_decode()` returns `null`.

### CODE REVIEW

File: src/ComposerRequireChecker/Cli/CheckCommand.php [hunk 1]
```diff
@@ -121,7 +121,7 @@ private function getCheckOptions(InputInterface $input) : Options
         }
 
         $jsonData = json_decode(file_get_contents($fileName), true);
-        if(false === $jsonData) {
+        if(false === $jsonData || null === $jsonData) {
```
Comment (Ocramius) [inline src/ComposerRequireChecker/Cli/CheckCommand.php]: A new test is needed to check this added condition

**Comment by villfa** [comment]
@Ocramius since there is no public method in _CheckCommand_ I had to move the condition in its own class to make it testable.

File: src/ComposerRequireChecker/JsonLoader.php [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    private $data;
+
+    /**
+     * @param string $path
+     */
+    public function __construct($path)
+    {
+        $this->path = $path;
+        if (!is_readable($path) || ($content = file_get_contents($path)) === false) {
+            $this->errorCode = self::ERROR_NO_READABLE;
```
Comment (Ocramius) [inline src/ComposerRequireChecker/JsonLoader.php]: Make it throw directly instead

File: src/ComposerRequireChecker/JsonLoader.php [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            return;
+        }
+        $this->data = json_decode($content, true);
+        if ($this->data === null && JSON_ERROR_NONE !== json_last_error()) {
+            $this->errorCode = self::ERROR_INVALID_JSON;
```
Comment (Ocramius) [inline src/ComposerRequireChecker/JsonLoader.php]: Make it throw directly instead

File: src/ComposerRequireChecker/JsonLoader.php [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            $this->errorMessage = json_last_error_msg();
+            return;
+        }
+        $this->errorCode = self::NO_ERROR;
+    }
+
+    /**
+     * @return string
+     */
+    public function getPath()
```
Comment (Ocramius) [inline src/ComposerRequireChecker/JsonLoader.php]: Not needed

File: src/ComposerRequireChecker/JsonLoader.php [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    {
+        return $this->path;
+    }
+
+    /**
+     * @return int
+     */
+    public function getErrorCode()
```
Comment (Ocramius) [inline src/ComposerRequireChecker/JsonLoader.php]: Not needed

File: src/ComposerRequireChecker/JsonLoader.php [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    {
+        return $this->errorCode;
+    }
+
+    /**
+     * @return string
+     */
+    public function getErrorMessage()
```
Comment (Ocramius) [inline src/ComposerRequireChecker/JsonLoader.php]: Not needed

File: test/ComposerRequireCheckerTest/JsonLoaderTest.php [hunk 7]
```diff
@@ -0,0 +1,40 @@
+<?php
+
+namespace ComposerRequireCheckerTest;
+
+use ComposerRequireChecker\JsonLoader;
+use PHPUnit\Framework\TestCase;
+
+class JsonLoaderTest extends TestCase
```
Comment (Ocramius) [inline test/ComposerRequireCheckerTest/JsonLoaderTest.php]: Add an `@covers` annotation

File: test/ComposerRequireCheckerTest/JsonLoaderTest.php [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    public function testHasErrorWithInvalidFile()
+    {
+        $path = __DIR__ . '/../fixtures/invalidJson';
+        $loader = new JsonLoader($path);
+        $this->assertEquals($loader->getErrorCode(), JsonLoader::ERROR_INVALID_JSON);
+        $this->assertEquals($loader->getPath(), $path);
+        $this->assertNotEmpty($loader->getErrorMessage());
+        $this->assertNull($loader->getData());
```
Comment (Ocramius) [inline test/ComposerRequireCheckerTest/JsonLoaderTest.php]: Just verify that an exception is thrown at construct in these tests ;-)

File: test/ComposerRequireCheckerTest/JsonLoaderTest.php [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+class JsonLoaderTest extends TestCase
+{
+
+    public function testHasErrorWithWrongPath()
+    {
+        $path = __DIR__ . '/wrong/path/non-existing-file.json';
+        $loader = new JsonLoader($path);
+        $this->assertEquals($loader->getErrorCode(), JsonLoader::ERROR_NO_READABLE);
+        $this->assertEquals($loader->getPath(), $path);
+        $this->assertNull($loader->getData());
```
Comment (Ocramius) [inline test/ComposerRequireCheckerTest/JsonLoaderTest.php]: Just verify that an exception is thrown at construct in these tests ;-)

File: test/ComposerRequireCheckerTest/JsonLoaderTest.php [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+use PHPUnit\Framework\TestCase;
+
+/**
+ * @covers \ComposerRequireChecker\JsonLoader
+ */
+class JsonLoaderTest extends TestCase
+{
+
+    /**
+     * @expectedException \ComposerRequireChecker\Exception\NotReadableException
```
Comment (Ocramius) [inline test/ComposerRequireCheckerTest/JsonLoaderTest.php]: Note that the `@expectedException` is to be avoided, while `->expectException()` should be placed right before the call causing the actual exception

File: test/ComposerRequireCheckerTest/JsonLoaderTest.php [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+     */
+    public function testHasErrorWithWrongPath()
+    {
+        $path = __DIR__ . '/wrong/path/non-existing-file.json';
+        new JsonLoader($path);
+    }
+
+    /**
+     * @expectedException \ComposerRequireChecker\Exception\InvalidJsonException
```
Comment (Ocramius) [inline test/ComposerRequireCheckerTest/JsonLoaderTest.php]: Same as above

File: src/ComposerRequireChecker/Cli/CheckCommand.php [hunk 12]
```diff
@@ -115,35 +115,19 @@ private function getCheckOptions(InputInterface $input) : Options
...
-
-        return new Options($jsonData);
-
+        return new Options((new JsonLoader($fileName))->getData());
     }
 
     /**
      * @param string $jsonFile
-     * @throws InvalidInputFileException
+     * @throws ComposerRequireChecker\Exception\InvalidJsonException
```
Comment (Ocramius) [inline src/ComposerRequireChecker/Cli/CheckCommand.php]: These are aliased incorrectly here: prepend a `\`

**Comment by villfa** [comment]
@Ocramius all your comments have been taken into account

**Comment by maglnet** [comment]
Thanks for the contribution and for the review!
Marking this as `@internal` sounds good, this gives the possibility to switch to some other library if needed.
I could possibly do this before the merge.

**Pull request merged**